### ISSUE TITLE
beancount cache should take modification date of plugins into account

### ISSUE THREAD

**Issue by tbm** [beancount/beancount#449]
I've been working on some beancount plugins.  It seems that beancount only looks at the beancount files to see if is has to refresh the cache.  When I modify the plugin, sometimes/often the new version isn't yet used.  I work around it by:

```
touch foo.beacount
bean-check foo.beancount
```

It seems beancount needs to look at the modification date of the plugins used by a beancount file to determine whether it can use the cache.

**Comment by blais** [comment]
Oh yeah, that's been annoying me directly for a while when developing.
I've been wanting to implement command-line options to disable or clear  the cache explicitly.

**Comment by blais** [comment]
You know about this, right?
https://github.com/beancount/beancount/blob/master/beancount/loader.py#L642

```python
    if os.getenv('BEANCOUNT_DISABLE_LOAD_CACHE') is None:
        _load_file = pickle_cache_function(
            os.getenv('BEANCOUNT_LOAD_CACHE_FILENAME') or PICKLE_CACHE_FILENAME,
            PICKLE_CACHE_THRESHOLD,
            _uncached_load_file)
```

**Comment by blais** [comment]
https://github.com/beancount/beancount/commit/5fa980176c7bd6fc9ae5d990c83867f942120f11

**Comment by tbm** [comment]
Thanks for adding these options, @blais.  I'm sure they'll be useful.

But I feel like you never answered my original question: why is the modification date of plugins not used to invalidate the cache?  The modification date on the beancount files is used, so why not also the date from the plugins that are used in those files?

**Comment by blais** [comment]
I think we could try to add that.

**Comment by blais** [comment]
Reopening to keep this alive. I can have a look next weekend.

**Comment by dnicolodi** [comment]
@tbm, computing the modification date of a plugin is far from trivial. We could use the modification date of the file implementing the plugin entry point, but what if the plugin is implemented in multiple files? We could track imports in the plugin (not easy but doable), but where do we stop? We could use some heuristic and consider imports from the same namespace of the plugin. But I am sure there would be cases in which this would not be sufficient. What about if a plugin uses some data files?

I think the only reasonable thing to do would be to have plugins export a version number and require the plugin developers to increment it when the plugin is updated. I don't know if this is much better than to require a manual cache invalidation.

**Comment by tbm** [comment]
You're making good points @dnicolodi. I didn't consider the case where modules consist of several files. I wasn't expecting beancount to look at all dependencies, although you make a fair point.

**Comment by blais** [comment]
Honestly I think just being able to explicitly call it with --no-cache (which also resets the cache) is sufficient for debugging plugins. I wouldn't bother going on a wild goose chase for timestamps down the import rabbit hole, not worth it.

**Comment by dnicolodi** [comment]
I agree with @blais : manually resetting the cache is sufficient for development. However, I think it would be nice for plugin authors to have a way to invalidate the cache when they distribute a plugin that introduces incompatible changes. I think having the plugin expose `__version__` along with the `__plugins__` list would nicely solve the problem. The collection of all `__version__` for the loaded plugins would need to be stored during load, saved alongside the cache (in the options dictionary?) and compared to the current one when the cache is loaded. I don't think it is too complex to do, but still some work...

**Comment by blais** [comment]
So doing "what's right" here yields such a marginal benefit I'm going to close this.
If you really care, send me a PR.

**Comment by dnicolodi** [comment]
I don't think there will be any activity soon on this. Can this issue bu "unpinned"?

**Comment by blais** [comment]
Note that we could be conservative and save the list and timestamp for all files under sys.modules. That would be one way to solve this issue easily.

**Comment by dnicolodi** [comment]
Timestamps of are not the best key on which to base cache invalidation. For example, Nix uses the same timestamp for all the installed files, thus this will not work there. I still think that an explicit `__version__` symbol exported by the plugins is the best way forward.

**Comment by blais** [comment]
Well it's really a change in timestamp that would trigger this. The point here is to trigger when a plugin changes, and that will be indicated by a timestamp. I don't see why it wouldn't work. We can probably try it, it would be 10 LOC or something.

**Comment by dnicolodi** [comment]
There are systems where the timestamps do not change. In Nix, for example, all installed files have the same modification time (corresponding to 1970-01-01 00:00:01, IIRC). Thus a package (or plugin) upgrade (that happens via the package manager) cannot be detected by looking ONLY at the modification time of the module file. What would work is looking at both path and mtime.

**Comment by blais** [comment]
Oh yes... but we don't care much about that. The only purpose of that would be to help in development, for people running from source.

**Comment by blais** [comment]
Reopening to hit this eventually, I think this can be tried
